From: Andreas Herrman <aherrman@de.ibm.com>
Subject: san_disc -c REPORT_LUNS results in Oops
Patch-Mainline: obsolete
References: 155752 - LTC22196

Description: statistics: oops on cleanup
Symptom:     NULL pointer dereference when calling
             statistic_interface_remove() or statistic_remove()
             respectively
Problem:     struct statistic is already gone when trying to use it
             in order access the parent interface\'s lock
Solution:    use a local variable for the interface pointer
Archs:       s390-31, s390-64 

Signed-off-by: Hannes Reinecke <hare@suse.de>

--- linux-2.6.16-rc5-git2-2/lib/statistic.c	2006-03-10 13:48:07.000000000 +0100
+++ linux/lib/statistic.c	2006-03-10 13:50:51.000000000 +0100
@@ -433,17 +433,20 @@ int statistic_create(struct statistic **
 int statistic_remove(struct statistic **stat_ptr)
 {
 	unsigned long flags;
+	struct statistic_interface *interface;
 
 	if (!*stat_ptr)
 		return -EINVAL;
 
-	statistic_lock((*stat_ptr)->interface, flags);
+	interface = (*stat_ptr)->interface;
+
+	statistic_lock(interface, flags);
 	if ((*stat_ptr)->release)
 		(*stat_ptr)->release(*stat_ptr);
 	list_del(&(*stat_ptr)->list);
 	kfree(*stat_ptr);
 	*stat_ptr = NULL;
-	statistic_unlock((*stat_ptr)->interface, flags);
+	statistic_unlock(interface, flags);
 
 	return 0;
 }
